language/move-prover/stackless-bytecode-generator/src/function_target_pipeline.rs

```diff
@@ -0,0 +1,89 @@
+// Copyright (c) The Libra Core Contributors
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::{
+    function_target::{FunctionTarget, FunctionTargetData},
+    stackless_bytecode_generator::StacklessBytecodeGenerator,
+};
+use itertools::Itertools;
+use spec_lang::env::{FunId, FunctionEnv, GlobalEnv, ModuleId};
+use std::collections::BTreeMap;
+
+/// A globally unique key to identify a function.
+type FunKey = (ModuleId, FunId);
+
+/// A data structure which holds data for multiple function targets, and allows to
+/// manipulate them as part of a transformation pipeline.
+#[derive(Debug, Default)]
+pub struct FunctionTargetsHolder {
+    targets: BTreeMap<FunKey, FunctionTargetData>,
+}
+
+/// A trait for processing a function target. Takes as parameter a target holder which can be
+/// mutated, the env of the function being processed, and the target data. During the time
+/// the processor is called, the target data is removed from the holder, and added back once
+/// transformation has finished. This allows the processor to take ownership
+/// on the target data. Notice that you can use `FunctionTarget{func_env, &data}` to temporarily
+/// construct a function target for access to the underlying data.
+pub trait FunctionTargetProcessor {
+    fn process(
+        &self,
+        targets: &mut FunctionTargetsHolder,
+        func_env: &FunctionEnv<'_>,
+        data: FunctionTargetData,
+    ) -> FunctionTargetData;
+}
+
+/// A processing pipeline for function targets.
+#[derive(Default)]
+pub struct FunctionTargetPipeline {
+    processors: Vec<Box<dyn FunctionTargetProcessor>>,
+}
+
+impl FunctionTargetsHolder {
+    /// Adds a new function target. The target will be initialized from the Move byte code.
+    pub fn add_target(&mut self, func_env: &FunctionEnv<'_>) {
+        let generator = StacklessBytecodeGenerator::new(func_env);
+        let data = generator.generate_function();
+        self.targets
+            .insert((func_env.module_env.get_id(), func_env.get_id()), data);
+    }
+
+    /// Gets a function target for read-only consumption.
+    pub fn get_target<'env>(&'env self, func_env: &'env FunctionEnv<'env>) -> FunctionTarget<'env> {
+        let data = self
+            .targets
+            .get(&(func_env.module_env.get_id(), func_env.get_id()))
+            .expect("function target exists");
+        FunctionTarget { func_env, data }
+    }
+
+    /// Processes the function target data for given function.
+    fn process(&mut self, func_env: &FunctionEnv<'_>, processor: &dyn FunctionTargetProcessor) {
+        let key = (func_env.module_env.get_id(), func_env.get_id());
+        let data = self.targets.remove(&key).expect("function target exists");
+        let data = processor.process(self, func_env, data);
```

Comment:
Nit: `let processed_data` or some other non-`data` name would be less confusing.

Reply:
Done, but frankly, I start liking Rust's scoping rules and do this now more and more often...

Reply:
It's a tough one--the shadowing/no-shadowing debate was constant in my old team's OCaml codebase as well. The compromise we eventually agreed on is that although shadowing is better for preventing errors (you never accidentally re-use a value!), it's harmful enough for readability (hard to see where things are bound via control-S or when skimming the code) that we agreed not to do it. 

In Rust, accidental reuse seems less likely (because so few values are `Copy` and you'll get a use-after-move on non-`Copy`), so my personal prefs lean toward no shadowing. Definitely a great debate though...